File: app/src/main/java/is/leap/android/sample/data/LeapSampleSharedPref.java
package is.leap.android.sample.data;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONObject;

import java.util.Set;

public class LeapSampleSharedPref {

    private static final String LEAP_SAMPLE_CACHE = "leap_sample_shared_pref";

    private static LeapSampleSharedPref INSTANCE;
    private final SharedPreferences sharedPreferences;

    public static final String APP_API_KEY = "apiKey";
    public static final String WEB_URL = "webUrl";

    private LeapSampleSharedPref(Context mContext) {
        sharedPreferences = mContext.getSharedPreferences(LEAP_SAMPLE_CACHE, Context.MODE_PRIVATE);
    }

    public static synchronized void init(Context mContext) {
        if (INSTANCE == null) INSTANCE = new LeapSampleSharedPref(mContext);
    }

    public static synchronized LeapSampleSharedPref getInstance() {
        if (INSTANCE == null) throw new RuntimeException("Make sure to call init at-least once.");
        return INSTANCE;
    }

    private void save(String key, JSONObject object) {
        save(key, object.toString());
    }

    private void save(String key, int value) {
        getEditor().putInt(key, value).apply();
    }

    private void save(String key, boolean value) {
        getEditor().putBoolean(key, value).apply();
    }

    public void save(String key, long value) {
        getEditor().putLong(key, value).apply();
    }

    public void save(String key, float value) {
        getEditor().putFloat(key, value).apply();
    }

    public void save(String key, Set<String> values) {
        getEditor().putStringSet(key, values).apply();
    }

    public void save(String key, String value) {
        getEditor().putString(key, value).apply();
    }

    // Remove & Clear methods
    public void remove(String key) {
        getEditor().remove(key).apply();
    }

    public void clear() {
        getEditor().clear().apply();
    }

    String getString(String key, String defaultValue) {
        try {
            return sharedPreferences.getString(key, defaultValue);
        } catch (Exception e) {
            return defaultValue;
        }
    }

    private SharedPreferences.Editor getEditor() {
        return sharedPreferences.edit();
    }

}

File: app/src/main/java/is/leap/android/sample/data/LeapSampleSharedPref.kt
package `is`.leap.android.sample.data

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONObject

class LeapSampleSharedPref private constructor(mContext: Context) {
    private val sharedPreferences: SharedPreferences
    private fun save(key: String, `object`: JSONObject) {
        save(key, `object`.toString())
    }

    private fun save(key: String, value: Int) {
        editor.putInt(key, value).apply()
    }

    private fun save(key: String, value: Boolean) {
        editor.putBoolean(key, value).apply()
    }

    fun save(key: String?, value: Long) {
        editor.putLong(key, value).apply()
    }

    fun save(key: String?, value: Float) {
        editor.putFloat(key, value).apply()
    }

    fun save(key: String?, values: Set<String?>?) {
        editor.putStringSet(key, values).apply()
    }

    fun save(key: String?, value: String?) {
        editor.putString(key, value).apply()
    }

    // Remove & Clear methods
    fun remove(key: String?) {
        editor.remove(key).apply()
    }

    fun clear() {
        editor.clear().apply()
    }

    private fun readString(key: String, defaultValue: String?): String? {
        return try {
            sharedPreferences.getString(key, defaultValue)
        } catch (e: Exception) {
            defaultValue
        }
    }

    private val editor: SharedPreferences.Editor
        private get() = sharedPreferences.edit()
    var apiKey: String?
        get() = readString(APP_API_KEY, null)
        set(apiKey) {
            save(APP_API_KEY, apiKey)
        }
    var webUrl: String?
        get() = readString(WEB_URL, null)
        set(webUrl) {
            save(WEB_URL, webUrl)
        }

    companion object {
        private const val LEAP_SAMPLE_CACHE = "leap_sample_shared_pref"
        private var INSTANCE: LeapSampleSharedPref? = null
        const val APP_API_KEY = "apiKey"
        const val WEB_URL = "webUrl"
        @Synchronized
        fun init(mContext: Context) {
            if (INSTANCE == null) INSTANCE = LeapSampleSharedPref(mContext)
        }

        @kotlin.jvm.JvmStatic
        @get:Synchronized
        val instance: LeapSampleSharedPref?
            get() {
                if (INSTANCE == null) throw RuntimeException("Make sure to call init at-least once.")
                return INSTANCE
            }
    }

    init {
        sharedPreferences = mContext.getSharedPreferences(LEAP_SAMPLE_CACHE, Context.MODE_PRIVATE)
    }
}